File: python/abc344/d.py
t = input()
n_t = len(t)
n = int(input())
a = []
for _ in range(n):
    _a = input().split()
    a.append(_a[1:])

ans = 200

dp = [200 for _ in range(n_t + 1)]
dp[0] = 0

for i in range(n_t):
    for j in range(n):
        for ij in range(len(a[j])):
            if i + len(a[j][ij]) <= n_t and t[n_t - i - len(a[j][ij]):n_t - i] == a[j][ij]:
                dp[i + len(a[j][ij])] = min(dp[i + len(a[j][ij])], dp[i] + 1)

print(dp[n_t] if dp[n_t] < 200 else -1)

File: rust/abc344/src/d.rs
#[allow(unused_imports)]
use proconio::{
    fastout, input, marker::Chars
};
use std::collections::VecDeque;


#[fastout]
fn main() {
	input! {
		t: Chars,
		n: usize,
	}
	let mut a = vec![];
	for _ in 0..n {
		input! {
			ai_n: usize,
			ai: [Chars; ai_n]
		}
		a.push(ai);
	}

	let n_t = t.len();
	let a_used = vec![vec![]; n];
	let mut ans = 200;

	let mut q = VecDeque::new();
	q.push_back((0, 0, a_used.clone()));

	while let Some((how_much, now, a_used)) = q.pop_front() {
		for i in 0..n {
			if a_used[i].len() != a[i].len() {
				for (j, a_j) in a[i].iter().enumerate() {
					if now + a_j.len() <= n_t && a_j.iter().eq(&t[n_t - now - a_j.len()..n_t - now]) {
						let mut _a_used = a_used.clone();
						_a_used[i].push(j);
						if now + a_j.len() < n_t {
							q.push_back((how_much + 1, now + a_j.len(), _a_used));
						} else {
							ans = ans.min(how_much + 1);
						}
					}
				}
			}
		}
	}

	println!("{}", if ans != 200 { ans } else { -1 });
}
